bmc/virtual_media.go

```diff
@@ -0,0 +1,70 @@
+package bmc
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/pkg/errors"
+)
+
+// VirtualMediaSetter controls the virtual media attached to a machine
+type VirtualMediaSetter interface {
+	SetVirtualMedia(ctx context.Context, kind string, mediaURL string) (ok bool, err error)
+}
+
+// VirtualMediaProviders is an internal struct to correlate an implementation/provider and its name
+type virtualMediaProviders struct {
+	name               string
+	virtualMediaSetter VirtualMediaSetter
+}
+
+// setVirtualMedia sets the virtual media.
+func setVirtualMedia(ctx context.Context, kind string, mediaURL string, b []virtualMediaProviders) (ok bool, metadata Metadata, err error) {
+	var metadataLocal Metadata
+Loop:
+	for _, elem := range b {
+		if elem.virtualMediaSetter == nil {
+			continue
+		}
+		select {
+		case <-ctx.Done():
+			err = multierror.Append(err, ctx.Err())
+			break Loop
+		default:
+			metadataLocal.ProvidersAttempted = append(metadataLocal.ProvidersAttempted, elem.name)
+			ok, setErr := elem.virtualMediaSetter.SetVirtualMedia(ctx, kind, mediaURL)
+			if setErr != nil {
+				err = multierror.Append(err, errors.WithMessagef(setErr, "provider: %v", elem.name))
+				continue
+			}
+			if !ok {
+				err = multierror.Append(err, fmt.Errorf("provider: %v, failed to set virtual media", elem.name))
+				continue
+			}
+			metadataLocal.SuccessfulProvider = elem.name
+			return ok, metadataLocal, nil
+		}
+	}
+	return ok, metadataLocal, multierror.Append(err, errors.New("failed to set virtual media"))
+}
+
+// SetVirtualMediaFromInterfaces identifies implementations of the virtualMediaSetter interface and passes the found implementations to the setVirtualMedia() wrapper
+func SetVirtualMediaFromInterfaces(ctx context.Context, kind string, mediaURL string, generic []interface{}) (ok bool, metadata Metadata, err error) {
+	bdSetters := make([]virtualMediaProviders, 0)
+	for _, elem := range generic {
+		temp := virtualMediaProviders{name: getProviderName(elem)}
+		switch p := elem.(type) {
```

Comment:
@joelrebel are we using go 1.18+? If so this could be simplified using generics, I think.

Reply:
yep we are on 1.18 I have to get round to giving generics a try for this use case.

Reply:
I just based this function on the other existing implementations. I'm not familiar with go's generics, so it might make more sense for someone else to convert all of them in a follow up PR?

Reply:
I'm not intimately familiar with the code base so I didn't realize there were a bunch of similar things. All good.